### PR TITLE
[WIP] Convert nested folders to use @cycle/collection

### PR DESCRIPTION
**Pull request by Widdershin**

### CODE REVIEW

**Comment by Widdershin** [comment]
@staltz fresh feedback would be great if you have the time.

You'll note that we've moved away from the sinkHandlers approach to having `Collection` take in a stream of add events and return a stream of collections.

This simplifies things a lot, but is also a bit more magical.

File: nested-folders/src/Folder.js [hunk 1]
```diff
@@ -67,37 +37,26 @@ function makeRandomColor() {
...
 
-    proxyChildAction$.imitate(childAction$.endWhen(removeSelf$.take(1)))
+    const vdom$ = Collection.pluck(children$, 'DOM')
+      .map(makeView(removable, color))
 
     return {
       DOM: vdom$,
-      action$,
-      id,
+      remove$
```
Comment (staltz) [inline nested-folders/src/Folder.js:55]: Trying to understand: where is the logic that deals with removal?

**Comment by staltz** [comment]
I liked the handlers/reducers approach. I feel like there are still different ways we could make an API for Collections, but I haven't spent enough time thinking about it.

File: nested-folders/src/Folder.js (hunk 1 above)
Comment (Widdershin) [inline nested-folders/src/Folder.js:55]: The signature for collection is now `Collection(component, sources, add$, removalSinkName = 'remove$')`.

So `Collection` internally constructs a stream of add/remove reducers and folds over it.

I'm still not sure how I feel about this new approach. It's less boilerplate, but it's also less explicit.

**Comment by Widdershin** [comment]
> I liked the handlers/reducers approach.

Good to know. I feel that there's value in the handlers/reducers approach, and perhaps we'll combine it with this style in a future version.

**Comment by Widdershin** [comment]
cc @hypnosphi

**Comment by Hypnosphi** [comment]
Which parts of it did you like specifically? I find those `.remove()` calls a bit imperative.

File: nested-folders/src/Folder.js [hunk 2]
```diff
@@ -67,37 +37,26 @@ function makeRandomColor() {
   return '#' + hexColor
 }
 
-function createFolderComponent({id, removable = true}) {
+function createFolderComponent({removable}) {
+  const ChildFolder = createFolderComponent({removable: true})
```
Comment (coupez) [inline nested-folders/src/Folder.js]: This will loop infinitely, no?

**Comment by staltz** [comment]
I liked the `.remove` API because it's immutable. Passive programming is the opposite of reactive, and we should by default choose reactive, but passive programming is suitable for data structures, which a collection is. And if it's immutable, like Immutable.js data structures, there's nothing imperative about it except the name.

**Comment by Hypnosphi** [comment]
Ok, but what are the other handler/reducer usecases, apart from removing the corresponding item? It looks to me that all the examples use the `handlers` argument only so:

``` js
{
  remove$: (collection, item) => collection.remove(items)
}
```

So making a shortcut for that reduces boilerplate and therefore makes sense. Maybe we should still leave the old `handlers` argument for other usecases, if we find such ones.

**Comment by staltz** [comment]
@Hypnosphi I don't know, maybe sort? Maybe replacing an instance?

File: nested-folders/src/Folder.js (hunk 2 above)
Comment (Widdershin) [inline nested-folders/src/Folder.js]: Ah woops, thought I tested that! Nice catch 😄 

Fixed by moving the declaration inside of `Folder`.

**Comment by Widdershin** [comment]
@Hypnosphi one of the things I can imagine handlers being useful for is when your parent component contains more state than just collection items. Say your state is `{items, removedItemCount}`. Your removal reducer might look like this:

``` js
(state, item) => ({items: state.items.remove(item), removedItemCount: state.removedItemCount + 1})
```

That's a contrived example but no doubt there are use cases out there like that.

The only thing for it in my opinion is to find more real world use cases, and attempt to cover them with the smallest possible API. Most of the examples for Cycle.js at the moment that use collections are relatively contrived.

**Comment by Hypnosphi** [comment]
Looks like folding on `Collection.merge(collection$, 'remove$').mapTo(1)`

**Comment by Widdershin** [comment]
Yep, but say you had another reducer that used the `removedItemCount` to change its behavior. That would not work with the `Collection.merge` style

**Comment by Widdershin** [comment]
Opened #28 to show @cycle/collection with the many example, including supporting a removal prop name that isn't `remove$`.

**Pull request closed without merging**